README
SMBIOS kernel module project
=============================

This project has the goal to make the SM-BIOS data visible in a human readable
format. Almost every computer with a BIOS does also provide System Management
BIOS data not there does not seem to exist a tool to easily read the management
information provided here in.

The System Management BIOS specifiaction is a standard provided by the DMTF
(distributed management task force, inc.) and available at
http://dmtf.org/standards/bios.php.

The smbios kernel module reads the SMBIOS management information and makes it
visible in three different formats:
1. /proc/smbios: the file names match the names of the SMBIOS types and the
files contain the interpreted (cooked) information of the corresponding type
2. /proc/smbios/raw:  the file names match the SMBIOS types and the contents
are in raw (binary) format, so any application can interpret the types on their
own
3. /proc/smbios/cooked: the file names match the SMBIOS types and the files
contain the interpreted (cooked) information of the corresponding type.

Currently, the cooked information does not yet cover each SMBIOS type nor does
it cover each entry in the SMBIOS types. This is subject to be done in the
future. The supported SMBIOS reference specification version is 2.3.1.

For installation instructions, please look at the INSTALL file.

The most useful stuff is already supported but there's still a lot to do.
Any help is appreciated, be it development support, reporting of bugs (or
patches) or simply by saying that you like it ;-)

Have fun,
Thomas


CHANGES
smbios project history

0.0.2 (2002-09-03)
  fixed a bug with 2.4 kernels (changed GFP_BUFFER to GFP_KERNEL)

0.0.1 (2001-09-15) 
	first release, a lot of things must be done....



main.c
/* Copyright (C) 2000-2001 Fujitsu Siemens Computers
   Joachim Braeuer
   This file is part of smbios

   smbios is free software; you can redistribute it and/or modify it
   under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License,
   or (at your option) any later version.

   smbios is distributed in the hope that it will be useful, but WITHOUT
   ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
   FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
   for more details.

   You should have received a copy of the GNU General Public License
   along with smbios; see the file COPYING. If not, write to the
   Free Software Foundation, Inc., 59 Temple Place - Suite 330,
   Boston, MA 02111-1307, USA.
*/

/* $Id: main.c,v 1.2 2002/09/03 10:33:12 bretthauert Exp $
 *
 * $Log: main.c,v $
 * Revision 1.2  2002/09/03 10:33:12  bretthauert
 * fixed a bug with 2.4 kernels (changed kmalloc parameter from
 * GFP_BUFFER to GFP_KERNEL
 *
 * Revision 1.1  2001/09/15 14:52:43  bretthauert
 * initial release
 *
 */

/** \file main.c
 *  kernel interface functions for smbios kernel module
 *
 *  \author Markus Lyra
 *  \author Thomas Bretthauer
 *  \date October 2000
 *  \version 0.1
 */
#ifndef __KERNEL__
#  define __KERNEL__
#endif
#ifndef MODULE
#  define MODULE
#endif

#define __NO_VERSION__		/* don't define kernel_verion in module.h */
#include <linux/module.h>
#include <linux/version.h>

char kernel_version[] = UTS_RELEASE;

#include <linux/kernel.h>	/* printk() */
#include <linux/errno.h>	/* error codes */
#include <linux/types.h>	/* size_t */
#include <linux/proc_fs.h>
#include <asm/io.h>		    /* ioremap() */

#include "strgdef.h"        /* holds all the interpreted/cooked string definitions */
#include "bios.h"		    /* local definitions */

EXPORT_NO_SYMBOLS;

/*
 *   Module stuff
 */

/** \fn int init_module (void)
 *  \brief module initialization
 *  \return 0 if successful initialization, an error code otherwise
 *
 *  \author Markus Lyra
 *  \author Thomas Bretthauer
 *  \author Joachim Braeuer
 *  \date Januar 2001
 */
int
init_module (void)
{
    int err = 0;


    PDEBUG ("starting module initialization\n");

    /*
     *  map the SMBIOS memory segment
     *
     *  ioremap (kernel) maps a physical address to a virtual address.
     *  bios start address is e.g. the Bios F-Segment
     */

    if (!(smbios_base = ioremap (BIOS_START_ADDRESS, BIOS_MAP_LENGTH)))
    {
        PDEBUG ("ioremap() for entry point failed\n");
        err = -ENXIO;
        goto ioremap_for_entry_point_failed;
    }

    PDEBUG ("BIOS base set to 0x%p\n", smbios_base);




    /*
     *   search for SM-BIOS or DMI-BIOS entry point
	 *		
	 * 	smbios_base holds the virtual starting address of the F-Segment.
	 *  first check for the smbios entry point. only if we can't find it
	 *	we look for the proprietary dmi bios entry point. This is a entry
	 *  point defined by Siemens Nixdorf before the 'real' interface has
	 *  been introduced.
     */

	/* smbios/dmibios_entry_point points to the beginning of the entry point
     * table. However, they are of different structure (real - proprietary)
     */
    if (!(smbios_entry_point = smbios_find_entry_point (smbios_base)))
    {
        PDEBUG ("SM-BIOS entry point not found\n");

        if (!(dmibios_entry_point = dmibios_find_entry_point (smbios_base)))
        {
	        PDEBUG ("DMI-BIOS entry point not found. Aborting...\n");
	        err = -ENXIO;
	        goto find_entry_point_failed;
        }
    }

    /*
     *  for SM-BIOS:
     *  check if Pointer to DMI structures exist.
     *  intermediate_string (_DMI_) is not '\0' terminated,
     *  so strncmp() with sizeof(DMI_STRING) - 1 is needed.
     */
    if (smbios_entry_point)
    {
        if (strncmp((char *) &(smbios_entry_point->intermediate_string),
						DMI_STRING, sizeof (DMI_STRING) - 1))
		{
	        PDEBUG ("Pointer to DMI structures not found!\n");
	        err = -ENXIO;
	        goto check_dmi_failed;
        }
    }

    /*
     *  map the SM-BIOS structures physical address range.
	 *  the 'real' smbios_structures_base contains the starting
	 *  address, where the instances of dmi structures are located.
     */
    if (smbios_entry_point)
    {
        if (!(smbios_structures_base =
	          ioremap (smbios_entry_point->struct_table_address,
		        (unsigned long) smbios_entry_point->struct_table_length)))
		{
	        PDEBUG ("ioremap() for structures table failed\n");
	        err = -ENXIO;
	        goto ioremap_for_structures_table_failed;
        }
    }

    /*
     * On the other hand, if we have the proprietary DMI Bios, smbios_structures_base
	 * contains a pointer to a table. This table contains an offset to the sm/dmi bios
     * structures for each single instance.
     */
    if (dmibios_entry_point)
    {
        if (!(smbios_structures_base = dmibios_entry_point->entry))
        {
	        PDEBUG ("invalid structure table entry%p,%p\n", smbios_structures_base,
		        dmibios_entry_point->entry);
	        err = -ENXIO;
	        goto ioremap_for_structures_table_failed;
        }
    }
    PDEBUG ("DMI structures base set to 0x%p\n", smbios_structures_base);



	/*
	 * assumption: we have one single pointer that points to the beginning of the
     * smbios/dmibios structures regardless what kind of bios we have.
     * (except DMIBios 2.0 - the interface we use was introduced with DMIBios 2.1)
     */


    /*
     *  create /proc entries
     */

    /* make /proc/smbios directory */
    if (!(smbios_proc_dir =
	       create_proc_entry (PROC_DIR_STRING, S_IFDIR, &proc_root)))
	{
        err = -ENOMEM;
        PDEBUG ("failed to create /proc/smbios directory entry\n");
        goto create_smbios_dir_failed;
    }
    PDEBUG ("/proc/smbios directory created.\n");
   
    /* make /proc/smbios/raw directory */
    if (!(smbios_raw_proc_dir =
				 create_proc_entry (PROC_DIR_STRING_RAW, S_IFDIR, smbios_proc_dir)))
	{
        err = -ENOMEM;
        PDEBUG ("failed to create /proc/smbios/raw directory entry\n");
        goto create_smbios_raw_dir_failed;
    }
    PDEBUG ("/proc/smbios/raw directory created.\n");

    /* make /proc/smbios/cooked directory */
    if (!(smbios_cooked_proc_dir =
				 create_proc_entry (PROC_DIR_STRING_COOKED, S_IFDIR, smbios_proc_dir)))
    {
        err = -ENOMEM;
		PDEBUG ("failed to create /proc/smbios/cooked directory entry\n");
		goto create_smbios_cooked_dir_failed;
    }
    PDEBUG ("/proc/smbios/cooked directory created.\n");

    /* create version file */
    if (smbios_entry_point)
    {
        if ((err = smbios_make_version_entry (smbios_proc_dir)))
	        goto smbios_make_version_entry_failed;
    }


   /* make raw files for SM-BIOS */
   if (smbios_entry_point)
   {
        if ((err = smbios_make_dir_entries (smbios_proc_dir, smbios_raw_proc_dir, smbios_cooked_proc_dir)))
	        goto make_smbios_dir_entries_failed;
   }
	
   /* make files for DMI-BIOS */
   if (dmibios_entry_point)
   {
        if ((err = dmibios_make_dir_entries (smbios_proc_dir, smbios_raw_proc_dir, smbios_cooked_proc_dir)))
	        goto make_smbios_dir_entries_failed;
   }

	

   PDEBUG ("module loaded succesfully\n");
   return 0;


/*
 * reached only if an error occured...
 *
 * whenever we need/create system resources and we run into an error, we clean up
 * in the oposite direction. Sample:
 *
 * allocate resource 1, 2, 3, 4, ...
 * error occurs
 * clean up resources ..., 4, 3, 2, 1
 *
 * the code above jumps into the correct position of the cleanup chain.
 */

make_smbios_dir_entries_failed:
    /* remove /proc/smbios/cooked files */
    smbios_destroy_dir_entries (smbios_cooked_proc_dir);
    /* remove /proc/smbios/raw files */
    smbios_destroy_dir_entries (smbios_raw_proc_dir);
    /* remove /proc/smbios files */
    smbios_destroy_dir_entries (smbios_proc_dir);
	
smbios_make_version_entry_failed:

    /* remove /proc/smbios/cooked directory */
    remove_proc_entry(PROC_DIR_STRING_COOKED, smbios_proc_dir);
	
create_smbios_cooked_dir_failed	:
    /* remove /proc/smbios/raw directory */
    remove_proc_entry(PROC_DIR_STRING_RAW, smbios_proc_dir);

create_smbios_raw_dir_failed:
    /* remove /proc/smbios directory */
    remove_proc_entry(PROC_DIR_STRING, &proc_root);
	
create_smbios_dir_failed:
    /* unmap the virtual to physical memory binding */
    if (smbios_entry_point)
        iounmap (smbios_structures_base);

ioremap_for_structures_table_failed:
check_dmi_failed:
find_entry_point_failed:
    /* unmap the virtual to physical memory binding */
    iounmap (smbios_base);

ioremap_for_entry_point_failed:
    return err;
}


/** \fn int cleanup_module (void)
 *  \brief module cleanup
 *
 *  \author Markus Lyra
 *  \author Thomas Bretthauer
 *  \date October 2000
 */
void
cleanup_module (void)
{
    /* remove /proc/smbios/cooked files */
    smbios_destroy_dir_entries (smbios_cooked_proc_dir);
    /* remove /proc/smbios/raw files */
    smbios_destroy_dir_entries (smbios_raw_proc_dir);
    /* remove /proc/smbios files */
    smbios_destroy_dir_entries (smbios_proc_dir);
	
    /* remove /proc/smbios/cooked directory */
    remove_proc_entry(PROC_DIR_STRING_COOKED, smbios_proc_dir);
    /* remove /proc/smbios/raw directory */
    remove_proc_entry(PROC_DIR_STRING_RAW, smbios_proc_dir);
    /* remove /proc/smbios directory */
    remove_proc_entry(PROC_DIR_STRING, &proc_root);
	
    /* unmap the virtual to physical memory binding */
    if (smbios_entry_point)
        iounmap (smbios_structures_base);

    /* unmap the virtual to physical memory binding */
    iounmap (smbios_base);

    PDEBUG ("module unloaded\n");
}


cooking.h
/* Copyright (C) 2001-2001 Fujitsu Siemens Computers
   Joachim Braeuer
   This file is part of smbios

   smbios is free software; you can redistribute it and/or modify it
   under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License,
   or (at your option) any later version.

   smbios is distributed in the hope that it will be useful, but WITHOUT
   ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
   FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
   for more details.

   You should have received a copy of the GNU General Public License
   along with smbios; see the file COPYING. If not, write to the
   Free Software Foundation, Inc., 59 Temple Place - Suite 330,
   Boston, MA 02111-1307, USA.
*/

/* $Id: cooking.h,v 1.2 2002/09/03 10:33:12 bretthauert Exp $
 *
 * $Log: cooking.h,v $
 * Revision 1.2  2002/09/03 10:33:12  bretthauert
 * fixed a bug with 2.4 kernels (changed kmalloc parameter from
 * GFP_BUFFER to GFP_KERNEL
 *
 * Revision 1.1  2001/09/15 14:52:43  bretthauert
 * initial release
 *
 */

/** \file cooking.h
 *  declarations and prototypes for interpreting DMI-BIOS and SM-BIOS types
 *
 *  \author Thomas Bretthauer
 *  \author Joachim Braeuer
 *  \version 0.1
 *  \date January 2001
 */
     
#ifndef __COOKING_H__
#define __COOKING_H__

/*
 * function prototypes
 */

/* description: see implementation file */
unsigned char * bios_cook (smbios_struct * smbiostype, unsigned int *length);

unsigned char * bios_cook_type_0 (smbios_struct * smbiostype, unsigned int *length);
unsigned char * bios_cook_type_1 (smbios_struct * smbiostype, unsigned int *length);
unsigned char * bios_cook_type_2 (smbios_struct * smbiostype, unsigned int *length);
unsigned char * bios_cook_type_3 (smbios_struct * smbiostype, unsigned int *length);
unsigned char * bios_cook_type_4 (smbios_struct * smbiostype, unsigned int *length);
unsigned char * bios_cook_type_5 (smbios_struct * smbiostype, unsigned int *length);
unsigned char * bios_cook_type_6 (smbios_struct * smbiostype, unsigned int *length);
unsigned char * bios_cook_type_7 (smbios_struct * smbiostype, unsigned int *length);
unsigned char * bios_cook_type_8 (smbios_struct * smbiostype, unsigned int *length);
unsigned char * bios_cook_type_9 (smbios_struct * smbiostype, unsigned int *length);
unsigned char * bios_cook_type_10 (smbios_struct * smbiostype, unsigned int *length);
unsigned char * bios_cook_type_11 (smbios_struct * smbiostype, unsigned int *length);
unsigned char * bios_cook_type_12 (smbios_struct * smbiostype, unsigned int *length);
unsigned char * bios_cook_type_13 (smbios_struct * smbiostype, unsigned int *length);
unsigned char * bios_cook_type_16 (smbios_struct * smbiostype, unsigned int *length);
unsigned char * bios_cook_type_17 (smbios_struct * smbiostype, unsigned int *length);
unsigned char * bios_cook_type_19 (smbios_struct * smbiostype, unsigned int *length);
unsigned char * bios_cook_type_20 (smbios_struct * smbiostype, unsigned int *length);
unsigned char * bios_cook_type_32 (smbios_struct * smbiostype, unsigned int *length);
unsigned char * bios_cook_type_127 (smbios_struct * smbiostype, unsigned int *length);



/**
 * cooked structures
 */

typedef struct smbios_header
{
    __u8    type                           __attribute__ ((packed));
    __u8    length                         __attribute__ ((packed));
    __u16   handle                         __attribute__ ((packed));
} smbios_header;

/* type 0 - bios information */
typedef struct smbios_type_0
{
	smbios_header	header;
    __u8	vendor							__attribute__ ((packed));
	__u8	version							__attribute__ ((packed));
	__u16	startaddr						__attribute__ ((packed));
	__u8	reldate							__attribute__ ((packed));
	__u8	romsize							__attribute__ ((packed));
	__u64	characteristics					__attribute__ ((packed));
    __u8	ext1							__attribute__ ((packed));
    __u8	ext2							__attribute__ ((packed));
} smbios_type_0;

/* type 1 - system information */
typedef struct smbios_type_1
{
	smbios_header	header;
    __u8	manufacturer					__attribute__ ((packed));
	__u8	productname						__attribute__ ((packed));
	__u8	version							__attribute__ ((packed));
	__u8	serialnumber					__attribute__ ((packed));
	__u8	uuid[16]						__attribute__ ((packed));
	__u8	wakeuptype						__attribute__ ((packed));
} smbios_type_1;

/* type 2 - base board information */
typedef struct smbios_type_2
{
	smbios_header	header;
    __u8	manufacturer					__attribute__ ((packed));
	__u8	product							__attribute__ ((packed));
	__u8	version							__attribute__ ((packed));
	__u8	serialnumber					__attribute__ ((packed));
} smbios_type_2;

/* type 3 - system enclosure information */
typedef struct smbios_type_3
{
	smbios_header	header;
    __u8	manufacturer					__attribute__ ((packed));
	__u8	type							__attribute__ ((packed));
	__u8	version							__attribute__ ((packed));
	__u8	serialnumber					__attribute__ ((packed));
	__u8	asset_tag						__attribute__ ((packed));
	__u8	bootup_state					__attribute__ ((packed));
	__u8	pwr_supply_state				__attribute__ ((packed));
	__u8	thermal_state					__attribute__ ((packed));
	__u8	security_state					__attribute__ ((packed));
	__u32	oem_defined						__attribute__ ((packed));
} smbios_type_3;

/* type 4 - processor information */
typedef struct smbios_type_4
{
	smbios_header	header;
    __u8	socket_designation				__attribute__ ((packed));
	__u8	type							__attribute__ ((packed));
	__u8	family							__attribute__ ((packed));
	__u8	manufacturer					__attribute__ ((packed));
	__u8	id[8]							__attribute__ ((packed));
	__u8	version							__attribute__ ((packed));
	__u8	voltage							__attribute__ ((packed));
	__u16	external_clock					__attribute__ ((packed));
	__u16	max_speed						__attribute__ ((packed));
	__u16	current_speed					__attribute__ ((packed));
	__u8	status							__attribute__ ((packed));
	__u8	upgrade							__attribute__ ((packed));
	__u16	l1_cache_handle					__attribute__ ((packed));
    __u16	l2_cache_handle					__attribute__ ((packed));
    __u16	l3_cache_handle					__attribute__ ((packed));
    __u8	serialnumber					__attribute__ ((packed));
    __u8	asset_tag						__attribute__ ((packed));
    __u8	part_number						__attribute__ ((packed));
} smbios_type_4;

/* type 5 - memory controller information - nowadays obsolete */
typedef struct smbios_type_5
{
	smbios_header	header;

    __u8	error_detection					__attribute__ ((packed));
	__u8	ecc								__attribute__ ((packed));
	__u8	supported_interleave			__attribute__ ((packed));
	__u8	current_interleave				__attribute__ ((packed));
	__u8	max_mem_module_size				__attribute__ ((packed));
	__u16	supported_speeds				__attribute__ ((packed));
	__u16	supported_types					__attribute__ ((packed));
	__u8	voltage							__attribute__ ((packed));
	__u8    nr_assoc_slots					__attribute__ ((packed));
} smbios_type_5;

/* type 6 - memory module information - nowadays obsolete */
typedef struct smbios_type_6
{
	smbios_header	header;

    __u8	designation						__attribute__ ((packed));
	__u8	connection						__attribute__ ((packed));
	__u8	current_speed					__attribute__ ((packed));
	__u16	current_type					__attribute__ ((packed));
	__u8	installed_size					__attribute__ ((packed));
	__u8	enabled_size					__attribute__ ((packed));
	__u8	status							__attribute__ ((packed));
} smbios_type_6;

/* type 7 - cache information */
typedef struct smbios_type_7
{
	smbios_header	header;

    __u8	designation						__attribute__ ((packed));
	__u16	configuration					__attribute__ ((packed));
	__u16	max_size	    				__attribute__ ((packed));
	__u16	installed_size					__attribute__ ((packed));
	__u16	supported_sram					__attribute__ ((packed));
	__u16	current_sram					__attribute__ ((packed));
	__u8	speed							__attribute__ ((packed));
	__u8	ecc_type		        		__attribute__ ((packed));
	__u8	cache_type						__attribute__ ((packed));
	__u8	associativity					__attribute__ ((packed));
} smbios_type_7;

/* type 8 - port connector information */
typedef struct smbios_type_8
{
	smbios_header	header;

    __u8	internal_designation			__attribute__ ((packed));
	__u8	internal_type					__attribute__ ((packed));
	__u8	external_designation	    	__attribute__ ((packed));
	__u8	external_type					__attribute__ ((packed));
	__u8	type					        __attribute__ ((packed));
} smbios_type_8;

/* type 9 - system slot information */
typedef struct smbios_type_9
{
	smbios_header	header;

    __u8	designation			            __attribute__ ((packed));
	__u8	type					        __attribute__ ((packed));
	__u8	width	    	                __attribute__ ((packed));
	__u8	usage					        __attribute__ ((packed));
	__u8	length					        __attribute__ ((packed));
	__u8	id1					            __attribute__ ((packed));
	__u8	id2					            __attribute__ ((packed));
	__u8	characteristics1				__attribute__ ((packed));
	__u8	characteristics2				__attribute__ ((packed));
} smbios_type_9;

/* type 10 - onboard devices */
/* due to the nature of the structure it is better to discard the
 * header for this structure
 */
typedef struct smbios_type_10
{
	/* do not use the header for this structure */
	/* smbios_header	header; */

	__u8	type					        __attribute__ ((packed));
	__u8	description    	                __attribute__ ((packed));
} smbios_type_10;

/* type 11 - oem strings */
typedef struct smbios_type_11
{
	smbios_header	header;

    __u8	nr_of_strings		            __attribute__ ((packed));
} smbios_type_11;

/* type 12 - system configuration options */
typedef struct smbios_type_12
{
	smbios_header	header;

    __u8	nr_of_strings		            __attribute__ ((packed));
} smbios_type_12;

/* type 13 - bios language information */
typedef struct smbios_type_13
{
	smbios_header	header;

    __u8	nr_of_strings   	            __attribute__ ((packed));
    __u8    flags                           __attribute__ ((packed));
    __u8    reserved[15]                    __attribute__ ((packed));
    __u8    current_language                __attribute__ ((packed));
} smbios_type_13;

/* type 16 - physical memory array */
typedef struct smbios_type_16
{
	smbios_header	header;

    __u8	location   	                    __attribute__ ((packed));
    __u8    use                             __attribute__ ((packed));
    __u8    ecc                             __attribute__ ((packed));
    __u32   max_capacity                    __attribute__ ((packed));
    __u16   error_handle                    __attribute__ ((packed));
    __u16   nr_of_devices                   __attribute__ ((packed));
} smbios_type_16;

/* type 17 - memory device */
typedef struct smbios_type_17
{
	smbios_header	header;

    __u16	memory_array_handle             __attribute__ ((packed));
    __u16   error_handle                    __attribute__ ((packed));
    __u16   total_width                     __attribute__ ((packed));
    __u16   data_width                      __attribute__ ((packed));
    __u16   size                            __attribute__ ((packed));
    __u8    form_factor                     __attribute__ ((packed));
    __u8    device_set                      __attribute__ ((packed));
    __u8    device_locator                  __attribute__ ((packed));
    __u8    bank_locator                    __attribute__ ((packed));
    __u8    mem_type                        __attribute__ ((packed));
    __u16   type_detail                     __attribute__ ((packed));
    __u16   speed                           __attribute__ ((packed));
    __u8    manufacturer                    __attribute__ ((packed));
    __u8    serial_number                   __attribute__ ((packed));
    __u8    asset_tag                       __attribute__ ((packed));
    __u8    part_number                     __attribute__ ((packed));
} smbios_type_17;

typedef struct smbios_type_19
{
	smbios_header	header;

    __u32	starting_adr                    __attribute__ ((packed));
    __u32   ending_adr                      __attribute__ ((packed));
    __u16   mem_array_handle                __attribute__ ((packed));
    __u8    partition_width                 __attribute__ ((packed));
} smbios_type_19;

typedef struct smbios_type_20
{
	smbios_header	header;

    __u32	starting_adr                    __attribute__ ((packed));
    __u32   ending_adr                      __attribute__ ((packed));
    __u16   mem_dev_handle                  __attribute__ ((packed));
    __u16   mem_array_mapped_adr_handle     __attribute__ ((packed));
    __u8    partition_row_position          __attribute__ ((packed));
    __u8    interleave_position             __attribute__ ((packed));
    __u8    interleaved_data_depth          __attribute__ ((packed));
} smbios_type_20;

typedef struct smbios_type_32
{
	smbios_header	header;

    __u8	reserved[6]                     __attribute__ ((packed));
    __u8    status                          __attribute__ ((packed));
} smbios_type_32;



/**
 * helper functions
 */

char * GetString(smbios_struct *structure, unsigned int stringnr);
unsigned int pow2(unsigned int);

#endif /* __COOKING_H__ */




bios.h
/* Copyright (C) 2001-2001 Fujitsu Siemens Computers
   Joachim Braeuer
   This file is part of smbios

   smbios is free software; you can redistribute it and/or modify it
   under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License,
   or (at your option) any later version.

   smbios is distributed in the hope that it will be useful, but WITHOUT
   ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
   FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
   for more details.

   You should have received a copy of the GNU General Public License
   along with smbios; see the file COPYING. If not, write to the
   Free Software Foundation, Inc., 59 Temple Place - Suite 330,
   Boston, MA 02111-1307, USA.
*/

/* $Id: bios.h,v 1.2 2002/09/03 10:33:12 bretthauert Exp $
 *
 * $Log: bios.h,v $
 * Revision 1.2  2002/09/03 10:33:12  bretthauert
 * fixed a bug with 2.4 kernels (changed kmalloc parameter from
 * GFP_BUFFER to GFP_KERNEL
 *
 * Revision 1.1  2001/09/15 14:52:43  bretthauert
 * initial release
 *
 */

/** \file bios.h
 *  declarations and prototypes of DMI-BIOS and SM-BIOS stuff
 *
 *  \author Markus Lyra
 *  \author Thomas Bretthauer
 *  \author Joachim Braeuer
 *  \version 0.11
 *  \date January 2001
 */
     
#ifndef __BIOS_H__
#define __BIOS_H__

/*
 *   Macros to help debugging
 */
/** to help debugging ... */
#undef PDEBUG /* undef it, just in case */
#ifdef _DEBUG_
#  define PDEBUG(fmt, args...) printk( KERN_DEBUG "smbios: " fmt, ## args)
#else
#  define PDEBUG(fmt, args...) /* not debugging: nothing */
#endif

/*
 *   Magic numbers
 */

/** start address of BIOS segment to scanned for SM-BIOS and DMI-BIOS */
#define BIOS_START_ADDRESS      0xF0000
/** length of the scanned BIOS area for SM-BIOS and DMI-BIOS */
#define BIOS_MAP_LENGTH         0x10000
/** magic 4 bytes to identify SM-BIOS entry point, paragraph boundary */
#define SMBIOS_MAGIC_DWORD      0x5F4D535F /* anchor string "_SM_" */
/** magic 4 bytes to identify DMI-BIOS entry point, byte boundary */
#define DMIBIOS_MAGIC_DWORD     0x494d445f /* anchor string "_DMI" */
/** identifier for SM-BIOS structures within SM-BIOS entry point */
#define DMI_STRING              "_DMI_"
/** list of types which are known to have subtyes; expandable! */
#define TYPES_WITH_SUBTYPES     185, 187, 208, 209, 210, 211, 212, 254
/** maximum block size for proc read function */
#define PROC_BLOCK_SIZE         (3*1024)


/** mode raw/cooked */
#define FILE_MODE_RAW       0
#define FILE_MODE_COOKED    1




/*
 *   Structures
 */

/** SM-BIOS entry point structure 
 * the SMBIOS Entry Point structure described below can be located by
 * application software by searching for the anchor string on paragraph
 * (16 byte) boundaries within the physical memory address range 000F0000h to
 * 000FFFFFh. This entry point encapsulates an intermediate anchor string
 * that is used by some existing DMI browsers.
 *
 * @note While the SMBIOS Major and Minor Versions (offsets 06h and 07h)
 * currently duplicate the information present in the SMBIOS BCD Revision
 * (offset 1Dh), they provide a path for future growth in this specification.
 * The BCD Revision, for example, provides only a single digit for each of
 * the major and minor version numbers.
 */
typedef struct smbios_entry_point_struct
{
	/** "_SM_", specified as four ASCII characters (5F 53 4D 5F) */
  __u32 anchor_string                  __attribute__ ((packed));
	/** checksum of the Entry Point Structure (EPS). This value, when added to 
	 * all other bytes in the EPS, will result in the value 00h (using 8 bit
	 * addition calculations). Values in the EPS are summed starting at offset
	 * 00h, for Entry Point Length bytes.*/
  __u8  entry_point_checksum           __attribute__ ((packed));
	/** Length of the Entry Point Structure, starting with the Anchor String 
	 * field, in bytes, currently 1Fh. */
  __u8  entry_point_length             __attribute__ ((packed));
	/** identifies the major version of this specification implemented in
	 * the table structures, e.g. the value will be 0Ah for revision 10.22
	 * and 02h for revision 2.1 */
  __u8  major_version                  __attribute__ ((packed));
	/** identifies the minor version of this specification implemented in
	 * the table structures, e.g. the value will be 16h for revision 10.22
	 * and 01h for revision 2.1 */
  __u8  minor_version                  __attribute__ ((packed));
	/** size of the largest SMBIOS structure, in bytes, and encompasses the
	 * structure's formatted area and text strings. This is the value returned
	 * as StructureSize from the Plug-n-Play 'Get SMBIOS Information' function */
	__u16 max_struct_size                __attribute__ ((packed));
	/** identifies the EPS revision implemented in this structure and identifies
	 * the formatting of offsets 0Bh to 0Fh, one of:
	 * 00h     Entry Point based on SMBIOS 2.1 definition; formatted area is
	 *         reserved and set to all 00h.
	 * 01h-FFh reserved for assignment via this specification */
  __u8  revision                       __attribute__ ((packed));
	/** the value present in the Entry Point Revision field defines the
	 * interpretation to be placed upon these5 bytes. */
  __u8  formated_area[5]               __attribute__ ((packed));
	/** "_DMI_", specified as five ASCII characters (5F 44 4D 49 5F) */
  __u8  intermediate_string[5]         __attribute__ ((packed));
	/** checksum of the Intermediate Entry Point Structure (IEPS). This value,
	 * when added to all other bytes in the IEPS, will result in the value 00h
	 * (using 8 bit addition calculations). Values in the IEPS are summed
	 * starting at offset 10h, for 0Fh bytes */
  __u8  intermediate_checksum          __attribute__ ((packed));
	/** the 32 bit physical starting address of the read-only SMBIOS Structure
	 * Table, that can start at any 32 bit address. This area contains all of the
	 * SMBIOS structures fully packed together. These structures can then be
	 * parsed to produce exactly the same format as that returned from a 'Get
	 * SMBIOS Structure' function call. */
  __u16 struct_table_length            __attribute__ ((packed));
  __u32 struct_table_address           __attribute__ ((packed));
  __u16 no_of_structures               __attribute__ ((packed));
  __u8  bcd_revision                   __attribute__ ((packed));
} smbios_entry_point_struct;

/** SM-BIOS and DMI-BIOS structure header */
typedef struct smbios_struct
{
  __u8  type                           __attribute__ ((packed));
  __u8  length                         __attribute__ ((packed));
  __u16 handle                         __attribute__ ((packed));
  __u8  subtype                        __attribute__ ((packed));
        /* ... other fields are structure dependend ... */
} smbios_struct;

/** DMI-BIOS structure header */
typedef struct dmibios_table_entry_struct
{
  __u16 size                           __attribute__ ((packed));
  __u16 handle                         __attribute__ ((packed));
  __u32 procedure                      __attribute__ ((packed));
} dmibios_table_entry_struct;

/** DMI-BIOS entry point structure */
typedef struct dmibios_entry_point_struct
{
  __u8  signature[10]                  __attribute__ ((packed));
  __u8  revision                       __attribute__ ((packed));
  dmibios_table_entry_struct entry[1]  __attribute__ ((packed));
} dmibios_entry_point_struct;


/*
 *   Variables
 */
extern struct proc_dir_entry * smbios_proc_dir;             /* /proc/smbios */
extern struct proc_dir_entry * smbios_raw_proc_dir;         /* /proc/smbios/raw */
extern struct proc_dir_entry * smbios_cooked_proc_dir;      /* /proc/smbios/cooked */
extern void * smbios_base;                                  /* F-Segment */
extern smbios_entry_point_struct * smbios_entry_point;      /* start of SMBIOS within the F-Segment */
extern dmibios_entry_point_struct * dmibios_entry_point;    /* start of DMIBIOS within the F-Segment */
extern void * smbios_structures_base;                       /* base of SMBIOS raw structures */
extern unsigned char smbios_types_with_subtypes[];
extern char smbios_version_string[32];                      /* e.g. V2.31 */

/*
 *   Functions
 */

/* for the description see the implementation file */
smbios_entry_point_struct * smbios_find_entry_point(void * base);
dmibios_entry_point_struct * dmibios_find_entry_point(void * base);
unsigned char smbios_check_entry_point(void * addr);

int smbios_type_has_subtype(unsigned char type);

int smbios_get_struct_length(smbios_struct * struct_ptr);
int dmibios_get_struct_length(smbios_struct * struct_ptr);

int smbios_version_proc (char *page, char **start, off_t off, int count, int *eof, void *data);
int bios_read_raw_proc(char * page, char ** start, off_t off, int count, int * eof, void * data);
int bios_read_cooked_proc(char *page, char **start, off_t off, int count, int *eof, void *data);
int dmibios_read_raw_proc(char * page, char ** start, off_t off, int count, int * eof, void * data);
int dmibios_read_cooked_proc(char *page, char **start, off_t off, int count, int *eof, void *data);

int smbios_make_dir_entries(struct proc_dir_entry *smbiosdir, struct proc_dir_entry *rawdir, struct proc_dir_entry *cookeddir);
int smbios_make_version_entry(struct proc_dir_entry *smbiosdir);
int dmibios_make_dir_entries(struct proc_dir_entry * smbiosdir, struct proc_dir_entry * rawdir, struct proc_dir_entry * cookeddir);

void smbios_destroy_dir_entries(struct proc_dir_entry * dir);

unsigned int smbios_get_readable_name_ext(char *readable_name, smbios_struct *struct_ptr);
unsigned int smbios_get_readable_name(char *readable_name, smbios_struct *struct_ptr);
int make_file_entries (char *filename, struct proc_dir_entry *dir, smbios_struct *struct_ptr, int mode);

#endif /* __BIOS_H__ */
